hwtracer/src/decode/ykpt/mod.rs

```diff
@@ -38,32 +80,151 @@ impl TraceDecoder for YkPTTraceDecoder {
+            if (hdr.flags() & libc::PF_W) == 0 {
+                let vaddr = usize::try_from(obj_base + hdr.vaddr()).unwrap();
+                let memsz = usize::try_from(hdr.memsz()).unwrap();
+                let key = vaddr..(vaddr + memsz);
+                segs.push((key.clone(), ()));
+            }
+        }
+    }
+    let tree = segs.into_iter().collect::<IntervalTree<usize, ()>>();
+    CodeSegs { tree }
+});
+
+const PT_MAX_COMPRETS: usize = 64;
+
+/// A data structure providing convenient access to virtual address ranges and memory slices for
+/// segments.
+///
+/// FIXME: For now this assumes that no dlopen()/dlclose() is happening.
+struct CodeSegs {
+    tree: IntervalTree<usize, ()>,
+}
+
+impl CodeSegs {
+    /// Obtain the virtual address range and a slice of memory for the segment containing the
+    /// specified virtual address.
+    fn seg<'s: 'a, 'a>(&'s self, vaddr: usize) -> Segment<'a> {
+        let mut hits = self.tree.query(vaddr..(vaddr + 1));
+        match hits.next() {
+            Some(x) => {
+                // Segments can't overlap.
+                debug_assert_eq!(hits.next(), None);
+
+                let slice = unsafe {
+                    slice::from_raw_parts(x.range.start as *const u8, x.range.end - x.range.start)
+                };
+                Segment {
+                    vaddrs: &x.range,
+                    slice,
+                }
+            }
+            None => todo!(), // Has an object been loaded or unloaded at runtime?
+        }
+    }
+}
+
+/// The virtual address range of and memory slice of one ELF segment.
+struct Segment<'a> {
+    /// The virtual address range of the segment.
+    vaddrs: &'a Range<usize>,
+    /// A memory slice for the segment.
+    slice: &'a [u8],
+}
+
 /// Represents a location in the instruction stream of the traced binary.
 #[derive(Eq, PartialEq)]
 enum ObjLoc {
-    /// A known location in the "main binary object" of the program.
+    /// A known byte offset in the "main binary object" of the program.
     MainObj(u64),
-    /// Either a location not in the "main binary object", or an unknown location.
-    OtherObjOrUnknown,
+    /// Anything else, as a virtual address (if known).
+    OtherObjOrUnknown(Option<usize>),
 }
 
 impl Debug for ObjLoc {
     fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
         match self {
-            Self::MainObj(v) => write!(f, "MainObj[0x{:x}]", v),
-            Self::OtherObjOrUnknown => write!(f, "OtherObjOrUnknown"),
+            Self::MainObj(v) => write!(f, "ObjLoc::MainObj(0x{:x})", v),
+            Self::OtherObjOrUnknown(e) => {
+                if let Some(e) = e {
+                    write!(f, "ObjLoc::OtherObjOrUnknown(0x{:x})", e)
+                } else {
+                    write!(f, "ObjLoc::OtherObjOrUnknown(???)")
+                }
+            }
         }
     }
 }
 
+/// The returns addresses that can appear on the compressed return stack.
```

Comment:
s/returns/return/

Reply:
Fixed in cc4209f889d54c8e30a1c48953453821052436c2